src/System.Text.JsonLab/System/Text/Json/JsonReader.cs

```diff
@@ -735,24 +735,20 @@ private void ConsumeSingleValue(ref BufferReader<byte> reader, byte marker)
             else if ((uint)(marker - '0') <= '9' - '0')
             {
                 ReadOnlySequence<byte> sequence = reader.Sequence.Slice(reader.Position);
-                Value = sequence.IsSingleSegment ? sequence.First.Span : sequence.ToArray();
+                Value = GetNumber(sequence.IsSingleSegment ? sequence.First.Span : sequence.ToArray());
```

Comment:
If it's not a single segment, we should first search the delimiter, and then allocate an array of the "right" size (i.e. fitting data up to the delimiter).

Reply:
Also, I wonder if it's worth, but maybe we could rent the array from a pool and let users return it to the same pool?

Reply:
> If it's not a single segment, we should first search the delimiter, and then allocate an array of the "right" size (i.e. fitting data up to the delimiter).

Since this code path is for single json values only, the delimiter will be at the end of the sequence anyway (if the json is valid), so there is no point of searching. The right size that fits the data is the entire length of the sequence, so calling ToArray is the same as finding the right size.

For json with multiple values (within objects/arrays), we use BufferReader.TryReadToAny which allocates the minimal amount necessary if the data straddles between two segments (we search for the delimiter there).

I will see if I can avoid the separate code path for single json values all together to make the code easier to read/understand. However, for valid json we are not over-allocating. It is possible, in the worst-case, for us to allocate quite a bit if the json is invalid.

Reply:
> Also, I wonder if it's worth, but maybe we could rent the array from a pool and let users return it to the same pool?

This requirement would need coordination with the BufferReader, since that is where the allocation is happening:
https://github.com/dotnet/corefxlab/blob/952ca00c3e5a6c930447cf53116f45b2aa007b67/src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs#L122

cc @JeremyKuhne

Reply:
I don't know that we have a good pattern yet for passing out rented memory without allocating. @KrzysztofCwalina is there a reason we don't have a struct that implements IMemoryOwner? Is that a rational thing? Or should there be some other pattern for return-when-done?